docs/topics/exceptions.md

```diff
@@ -1,122 +1,706 @@
 
-* [Java's checked exceptions were a mistake](https://radio-weblogs.com/0122027/stories/2003/04/01/JavasCheckedExceptionsWereAMistake.html) (Rod Waldhoff)
-* [The Trouble with Checked Exceptions](https://www.artima.com/intv/handcuffs.html) (Anders Hejlsberg)
+To create one, you can define a class that extends `Exception`:
 
-If you want to alert callers about possible exceptions when calling Kotlin code from Java, Swift, or Objective-C,
-you can use the `@Throws` annotation. Read more about using this annotation [for Java](java-to-kotlin-interop.md#checked-exceptions)
-and [for Swift and Objective-C](native-objc-interop.md#errors-and-exceptions).
+```kotlin
+class MyException: Exception("My message")
+```
 
-## The Nothing type
+In this example, there is a default error message, "My message", but you can leave it blank if you want.
+
+Custom exceptions can also be a subclass of any pre-existent exception subclass, like the `ArithmeticException` subclass:
 
-`throw` is an expression in Kotlin, so you can use it, for example, as part of an Elvis expression:
+```kotlin
+class NumberTooLargeException: ArithmeticException("My message")
+```
+
+> Exceptions in Kotlin are stateful objects, carrying information specific to the context of their creation, referred to as the [stack trace](#stack-trace).
+> Avoid creating exceptions using `object` declarations. 
+> Instead, create a new instance of the exception every time you need one. 
+> This way, you can ensure the exception's state accurately reflects the specific context.
+> 
+{type="tip"}
+
+> In Kotlin, [classes are final by default](inheritance.md) and cannot be subclassed unless the parent
+> class is declared as `open`.
+> 
+> For example:
+>
+> ```kotlin
+> // Declares a custom exception as an open class making it subclassable
+> open class MyCustomException(message: String): Exception(message)
+>
+> // Creates a subclass of the custom exception
+> class SpecificCustomException: MyCustomException("Specific error message")
+> ```
+>
+{type="note"}
+
+Custom exceptions function just like built-in exceptions. You can throw them using the `throw` keyword, 
+and handle them with `try-catch-finally` blocks. Let's look at an example to demonstrate:
 
 ```kotlin
-val s = person.name ?: throw IllegalArgumentException("Name required")
+class NegativeNumberException: Exception("Parameter is less than zero.")
+class NonNegativeNumberException: Exception("Parameter is a non-negative number.")
+
+fun myFunction(number: Int) {
+    if (number < 0) throw NegativeNumberException()
+    else if (number >= 0) throw NonNegativeNumberException()
+}
+
+fun main() {
+    
+    // Change the value in this function to a get a different exception
+    myFunction(1)
+}
 ```
+{kotlin-runnable="true"}
+
+In complex applications with diverse error scenarios,
+creating a hierarchy of exceptions can help making the code clearer and more specific.
```

Comment:
hmm 🤔 it's more concise but to me it reads a bit strange. 

In applications with diverse error scenarios, an exception hierarchy makes code clearer and more specific. 

But this way I think we lose the additional information that it CAN make it more specific - not necessarily, but usually it would

I think I prefer the original the most here.

Reply:
Ok, up to you